datatrails: interpolate adhoc variables and datasource variables when opening "explore metrics" from dashboard panels

darrenjaneczek opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**Which issue(s) does this PR fix?**:

Bug:
- Datasource variables and adhoc filter variables were being ignored when opening explore metrics from dashboard panels

Current behavior matches the dashboard behavior when it comes to interpolation:
- **(A)** Datasource variables will always be interpolated
- **(B)** AdHoc filters will be included when the datasource of the query matches the datasource of the adhoc filter variable
- **(C)** In scenes-based dashboards, the adhoc variable will be interpolated if it is referenced
  - This is consistent with panel query behavior in scenes-based dashboards
- **(D)** ~Legacy dashboards will not interpolate the adhoc variable (Except for case B)~
  - ~This is consistent with dashboard panel behavior for legacy dashboards.~
  - Support for legacy-based dashboards will simply be dropped.

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

darrenjaneczek committed:
fix: interpolate "explore metrics" from panels

torkelo commented on public/app/features/trails/Integrations/interpolation.ts:
```diff
@@ -0,0 +1,56 @@
+import { DashboardModel } from 'app/features/dashboard/state';
+import { DashboardScene } from 'app/features/dashboard-scene/scene/DashboardScene';
+
+export function interpolateVariables(
```
you can replace all this implementation ds.interpolateVariablesInQueries  (this is a function on the datasource instance). 

This what we do when we go from dashboard to Explore or from data trails to Explore

https://github.com/grafana/grafana/blob/main/public/app/core/utils/explore.ts#L81

torkelo commented on public/app/features/trails/Integrations/interpolation.ts:
```diff
@@ -0,0 +1,56 @@
+  dsInstanceSettings: DataSourceInstanceSettings,
+  queries: DataQuery[]
+) {
+  const promDataSourceForUtility = new PrometheusDatasource({ ...dsInstanceSettings, jsonData: {} });
```
why new up a new data source?

torkelo replied:
you can get the query data source using `await getDataSourceSrv().get(query.datasource);`

darrenjaneczek replied:
For some reason, I had it in my head that I was still in the 'thread' that was generating the menu items and assumes "await" was off limits. Thanks for reminding me that I am within the callback post user selection, so I can "await" all I like.

darrenjaneczek replied:
Hmm... wait, I am still in the thread that generates the menu items.
This means that my labels for the menu items will have to use the variable names, but the execution of the action can take advantage of the interpolation... Unless I keep this weird strategy of constructing the datasource just in time... Let me see...

darrenjaneczek replied:
@torkelo 
> why new up a new data source?

Though it is awkward, it does allow me to take advantage of the utility functions in the prom datasource for interpolating variables. By recreating the datasource, I can call them within the context of generating these menu items, without `await`, and use the interpolated queries to parse for metrics/filters to generate labelled menu items and corresponding click handlers.

Without interpolating in advance:
- I cannot label the menu items with the interpolated values
- In certain cases I cannot properly identify prometheus metrics and filters due to the language parser not being to handle non-interpolated variable notation

I tried a few things, but the menu list that we populate doesn't tolerate changes. I don't expect there would be a way to make it optionally operate on `Promise<MenuItem>`, or some kind of refresh callback.

torkelo replied:
I could make this function async without any changes (in scenes version) as addDataTrailPanelAction was called from an async function. 

newing up PrometheusDataSource could be problematic (esp as the team wants to externalize this data source) so it would be loaded async as any plugin.

torkelo commented on public/app/features/trails/Integrations/interpolation.ts:
```diff
@@ -0,0 +1,56 @@
+) {
+  const promDataSourceForUtility = new PrometheusDatasource({ ...dsInstanceSettings, jsonData: {} });
+
+  return queries.filter(isPromQuery).map((query) => {
+    const queryDataSourceUid = getInterpolatedDataSourceUid(query?.datasource);
+    return promDataSourceForUtility.applyTemplateVariables(query, {}, getAdHocFilters(dashboard, queryDataSourceUid));
+  });
+}
+
+function getAdHocFilters(dashboard: DashboardModel | DashboardScene, queryDataSourceUid: string | undefined) {
+  const adhocFilters: AdHocVariableFilter[] = [];
+  if (dashboard instanceof DashboardModel) {
+    dashboard.getVariables().forEach((variable) => {
+      if (variable.type === 'adhoc') {
+        const variableDataSourceUid = getInterpolatedDataSourceUid(variable.datasource);
+        if (variableDataSourceUid === queryDataSourceUid) {
+          variable.filters.forEach((filter) => adhocFilters.push(filter));
+        }
+      }
+    });
+  } else {
+    dashboard.state.$variables?.state.variables.filter(isAdHocFiltersVariable).forEach((variable) => {
+      const variableDataSourceUid = getInterpolatedDataSourceUid(variable.state.datasource);
+      if (variableDataSourceUid === queryDataSourceUid) {
+        variable.state.filters.forEach((filter) => adhocFilters.push(filter));
+      }
+    });
+  }
+
+  return adhocFilters;
+}
```
This wont work in scenes, 

Implemented a simple scenes version here: https://github.com/grafana/grafana/pull/86779

I think we can skip support for old dashboards, it's going away soon anyway and this use case from dashboards is a bit experimental anyway

darrenjaneczek replied:
It would certainly simplify things to skip support for the old dashboards.
Are we expecting the change to scenes dashboards to make it into 11.0.x?

darrenjaneczek replied:
Dropped support for legacy dashboards and took advantage of the async and scenes objects.
Everything feels *much* cleaner now.

torkelo committed:
DashboardScene: Alternative to interpolating variables (#86779)

* DashboardScene: Alternative to interpolating variables

* fix: remove

---------

Co-authored-by: Darren Janeczek <darren.janeczek@grafana.com>

darrenjaneczek committed:
fix: remove support for legacy dashboard

- simplify code
- take advantage of scenes dashboard async and datasource api object
